runtime/memprof.c

```diff
@@ -15,6 +15,276 @@
+  domain->config.tracker = Val_unit;
+
+  /* create initial thread for domain */
+  memprof_thread_t thread = thread_create(domain);
+  if (thread) {
+    domain->current = thread;
+  } else {
+    domain_destroy(domain);
+    domain = NULL;
+  }
+  return domain;
+}
+
+/**** Interface to domain module ***/
+
+void caml_memprof_new_domain(caml_domain_state *parent,
+                             caml_domain_state *child)
+{
+  memprof_domain_t domain = domain_create(child);
+
+  child->memprof = domain;
+  /* if parent domain is profiling, child domain should also be profiling */
+  if (domain && parent && parent->memprof
+      && !atomic_load(&parent->memprof->config.stopped)) {
+    domain->config = parent->memprof->config;
+  }
+}
+
+void caml_memprof_delete_domain(caml_domain_state *domain)
+{
+  if (!domain->memprof) {
+    return;
+  }
+  domain_destroy(domain->memprof);
+  domain->memprof = NULL;
+}
+
+/**** Interface with domain action-pending flag ****/
+
+/* If profiling is active in the current domain, and we may have some
+ * callbacks pending, set the action pending flag. */
+
+static void set_action_pending_as_needed(memprof_domain_t domain)
+{
+  /* if (condition) caml_set_action_pending(domain->caml_state); */
+}
+
+/* Set the suspended flag on `domain` to `s`. */
+
+static void update_suspended(memprof_domain_t domain, bool s)
+{
+  if (domain->current) {
+    domain->current->suspended = s;
+  }
+  caml_memprof_renew_minor_sample(domain->caml_state);
+  if (!s) set_action_pending_as_needed(domain);
+}
+
+/* Set the suspended flag on the current domain to `s`. */
+
+void caml_memprof_update_suspended(bool s) {
+  update_suspended(Caml_state->memprof, s);
+}
+
+/**** Sampling procedures ****/
+
+Caml_inline bool running(memprof_domain_t domain)
+{
+  memprof_thread_t thread = domain->current;
+
+  return (thread
+          && !atomic_load(&domain->config.stopped)
+          && domain->config.lambda > 0
+          && !thread->suspended);
+}
+
+/* Renew the next sample in the minor heap. This needs to be called
+ * after each minor sample and after each minor collection. In
+ * practice, this is called at each minor sample, at each minor
+ * collection, and when sampling is suspended and unsuspended. Extra
+ * calls do not change the statistical properties of the sampling
+ * because of the memorylessness of the geometric distribution. */
+
+void caml_memprof_renew_minor_sample(caml_domain_state *state)
+{
+  memprof_domain_t domain = state->memprof;
+  value *trigger = state->young_start;
+  if (running(domain)) {
+    /* set trigger based on geometric distribution */
+  }
+  CAMLassert((trigger >= state->young_start) &&
+             (trigger < state->young_ptr));
+  state->memprof_young_trigger = trigger;
+  caml_reset_young_limit(state);
+}
+
+/**** Interface with systhread. ****/
+
+CAMLexport memprof_thread_t caml_memprof_new_thread(caml_domain_state *domain)
+{
+  CAMLassert(domain->memprof);
```

Comment:
I agree (although I'm not sure what happens when a domain-startup allocation fails; it's not clear from a brief check of domain.c).

Reply:
I've looked into this a bit more deeply and rearranged some of the code in `domain_create` so that we don't set `domain_self` if either this or the `caml_domain_state` allocation fails, which both callers of that function now turn into a `caml_fatal_error`. That is at least consistent; possibly `Domain.spawn` should raise an exception rather than a fatal error, but there are other ways for that to get `caml_fatal_error` (e.g. "Failed to create domain: signal stack") so that is also more consistent than previously.

Reply:
Good!